// posthog-js/lib/src/extensions/exception-autocapture/index.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { assignableWindow, window } from '../../utils/globals';
import { createLogger } from '../../utils/logger';
import { EXCEPTION_CAPTURE_ENABLED_SERVER_SIDE } from '../../constants';
import { isObject, isUndefined } from '../../utils/type-utils';
var logger = createLogger('[ExceptionAutocapture]');
var ExceptionObserver = /** @class */ (function () {
    function ExceptionObserver(instance) {
        var _this = this;
        var _a;
        this.startCapturing = function () {
            var _a, _b, _c, _d, _e, _f;
            if (!window || !_this.isEnabled) {
                return;
            }
            var wrapOnError = (_b = (_a = assignableWindow.__PosthogExtensions__) === null || _a === void 0 ? void 0 : _a.errorWrappingFunctions) === null || _b === void 0 ? void 0 : _b.wrapOnError;
            var wrapUnhandledRejection = (_d = (_c = assignableWindow.__PosthogExtensions__) === null || _c === void 0 ? void 0 : _c.errorWrappingFunctions) === null || _d === void 0 ? void 0 : _d.wrapUnhandledRejection;
            var wrapConsoleError = (_f = (_e = assignableWindow.__PosthogExtensions__) === null || _e === void 0 ? void 0 : _e.errorWrappingFunctions) === null || _f === void 0 ? void 0 : _f.wrapConsoleError;
            if (!wrapOnError || !wrapUnhandledRejection || !wrapConsoleError) {
                logger.error('failed to load error wrapping functions - cannot start');
                return;
            }
            try {
                if (!_this.unwrapOnError) {
                    _this.unwrapOnError = wrapOnError(_this.captureException.bind(_this));
                }
                if (!_this.unwrapUnhandledRejection) {
                    _this.unwrapUnhandledRejection = wrapUnhandledRejection(_this.captureException.bind(_this));
                }
                if (!_this.unwrapConsoleError) {
                    _this.unwrapConsoleError = wrapConsoleError(_this.captureException.bind(_this));
                }
            }
            catch (e) {
                logger.error('failed to start', e);
                _this.stopCapturing();
            }
        };
        this.instance = instance;
        this.remoteEnabled = !!((_a = this.instance.persistence) === null || _a === void 0 ? void 0 : _a.props[EXCEPTION_CAPTURE_ENABLED_SERVER_SIDE]);
        this.config = this.requiredConfig();
        this.startIfEnabled();
    }
    ExceptionObserver.prototype.requiredConfig = function () {
        var providedConfig = this.instance.config.capture_exceptions;
        var config = {
            capture_unhandled_errors: false,
            capture_unhandled_rejections: false,
            capture_console_errors: false,
        };
        if (isObject(providedConfig)) {
            config = __assign(__assign({}, config), providedConfig);
        }
        else if (isUndefined(providedConfig) ? this.remoteEnabled : providedConfig) {
            config = __assign(__assign({}, config), { capture_unhandled_errors: true, capture_unhandled_rejections: true });
        }
        return config;
    };
    Object.defineProperty(ExceptionObserver.prototype, "isEnabled", {
        get: function () {
            return (this.config.capture_console_errors ||
                this.config.capture_unhandled_errors ||
                this.config.capture_unhandled_rejections);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ExceptionObserver.prototype, "hasHandlers", {
        get: function () {
            return (!isUndefined(this.unwrapOnError) &&
                !isUndefined(this.unwrapUnhandledRejection) &&
                !isUndefined(this.unwrapConsoleError));
        },
        enumerable: false,
        configurable: true
    });
    ExceptionObserver.prototype.startIfEnabled = function () {
        if (this.isEnabled && !this.hasHandlers) {
            logger.info('enabled, starting...');
            this.loadScript(this.startCapturing);
        }
    };
    ExceptionObserver.prototype.loadScript = function (cb) {
        var _a, _b;
        if (this.hasHandlers) {
            // already loaded
            cb();
        }
        (_b = (_a = assignableWindow.__PosthogExtensions__) === null || _a === void 0 ? void 0 : _a.loadExternalDependency) === null || _b === void 0 ? void 0 : _b.call(_a, this.instance, 'exception-autocapture', function (err) {
            if (err) {
                return logger.error('failed to load script', err);
            }
            cb();
        });
    };
    ExceptionObserver.prototype.stopCapturing = function () {
        var _a, _b, _c;
        (_a = this.unwrapOnError) === null || _a === void 0 ? void 0 : _a.call(this);
        this.unwrapOnError = undefined;
        (_b = this.unwrapUnhandledRejection) === null || _b === void 0 ? void 0 : _b.call(this);
        this.unwrapUnhandledRejection = undefined;
        (_c = this.unwrapConsoleError) === null || _c === void 0 ? void 0 : _c.call(this);
        this.unwrapConsoleError = undefined;
    };
    ExceptionObserver.prototype.onRemoteConfig = function (response) {
        var _a;
        var autocaptureExceptionsResponse = response.autocaptureExceptions;
        // store this in-memory in case persistence is disabled
        this.remoteEnabled = !!autocaptureExceptionsResponse || false;
        this.config = this.requiredConfig();
        if (this.instance.persistence) {
            this.instance.persistence.register((_a = {},
                _a[EXCEPTION_CAPTURE_ENABLED_SERVER_SIDE] = this.remoteEnabled,
                _a));
        }
        this.startIfEnabled();
    };
    ExceptionObserver.prototype.captureException = function (errorProperties) {
        var posthogHost = this.instance.requestRouter.endpointFor('ui');
        errorProperties.$exception_personURL = "".concat(posthogHost, "/project/").concat(this.instance.config.token, "/person/").concat(this.instance.get_distinct_id());
        this.instance.exceptions.sendExceptionEvent(errorProperties);
    };
    return ExceptionObserver;
}());
export { ExceptionObserver };
